File: Shells/Java/src/cspSolver/Trail.java
package cspSolver;

import java.util.EmptyStackException;
import java.util.Stack;

/**
 * Represents the trail of changes made. This allows backtracking to occur. 
 * Trail follows the singleton design pattern. 
 */
public class Trail {
	
	//===============================================================================
	// Properties
	//===============================================================================
	
	private static Stack<Object[]> trail = new Stack<Object[]>();
	private static final Trail TRAIL = new Trail();
	private static Stack<Integer> breadcrumbs = new Stack<Integer>();
	private Trail()
	{}
	
	/**
	 * returns the trail.
	 * @return
	 */
	public static Trail getTrail()
	{
		return TRAIL;
	}
	
	public static void clearTrail()
	{
		trail.clear();
	}
	//===============================================================================
	// Accessors
	//===============================================================================
	
	public int size()
	{
		return trail.size();
	}
	
	//===============================================================================
	// Modifiers
	//===============================================================================
	
	/**
	 * places a marker at the current point in the trail. Each time undo is called, the 
	 * latest marker is popped and the trail 
	 */
	public void placeBreadCrumb()
	{
		breadcrumbs.add(trail.size());
	}
	/**
	 * Adds a deep copy of a variable and its domain onto the trail.
	 * @param v Variable to copy onto the trail
	 */
	public void push(Variable v)
	{
		Object[] vPair = { v, new Domain(v.getDomain())};
		trail.push(vPair);
	}
	
	/**
	 * Pops changes pushed onto the trail until it reaches the latest marker. 
	 * Also pops the latest marker. 
	 * @param targetSize target position on the trail to backtrack to
	 * @throws EmptyStackException if trail is empty
	 */
	public void undo() throws EmptyStackException 
	{
		int targetSize = breadcrumbs.pop();
		for (int size = trail.size(); size > targetSize; size--)
		{
			Object[] vPair = trail.pop();
			Variable v = (Variable) vPair[0];
			v.setDomain((Domain) vPair[1]);
		}
	}
}

File: Shells/Python/trail.py
# import copy
import variable
import domain
import copy

class Trail:
    def __init__(self):
        """
            Represents the trail of changes made. This allows backtracking to occur.
        """
        self.trailStack = []
        self.trailMarker = []

    ######### Accessors Method #########
    def size(self):
        return len(self.trailStack)

    ######### Modified Method #########
    def placeTrailMarker(self):
        """
            places a marker at the current point in the trail. Each time undo is
            called, the latest marker is popped and the trail
        """
        self.trailMarker.append(len(self.trailStack))
        # print("self.trailMarker: " + str(self.trailMarker))

    def push(self, v):
        """
            Adds a deep copy of a variable and its domain onto the trail.
        """
        # print("v in push--> " + str(v))
        domainCopy = copy.deepcopy(v.domain)
        vPair = [v, domainCopy] # v Variable and its deepcopy backup domain onto the trail
        self.trailStack.append(vPair)

        # print("======================= ")
        # print("self.trailStack at the moment: ")
        # for i in self.trailStack:
        #     print("variable --> " + str(i[0]))
        #     print("domain backup --> " + str(i[1]))
        # print("======================= ")

    def undo(self):
        """
            Pops changes pushed onto the trail until it reaches the latest marker.
            Also pops the latest marker.
        """
        targetSize = self.trailMarker.pop() # targetSize target position on the trail to backtrack to
        size = len(self.trailStack)
        while size > targetSize:
            vPair = self.trailStack.pop()
            v = vPair[0]
            v.setDomain(vPair[1])
            size -= 1

    def __str__(self):
        output = "trailStack: " + str(self.trailStack)
        output += "\ntrailMarker: " + str(self.trailMarker)
        return output

# Trail follows the singleton design pattern. A single class/object ever created.
masterTrailVariable = Trail()
